rust-runtime/aws-smithy-runtime-api/src/client.rs

```diff
@@ -3,6 +3,68 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+macro_rules! new_type_future {
+    (
+        doc = $type_docs:literal,
+        pub struct $future_name:ident<$output:ty, $err:ty>,
+    ) => {
+        pin_project_lite::pin_project! {
+            #[allow(clippy::type_complexity)]
+            #[doc = $type_docs]
+            pub struct $future_name {
+                #[pin]
+                inner: aws_smithy_async::future::now_or_later::NowOrLater<
+                    Result<$output, $err>,
+                    aws_smithy_async::future::BoxFuture<$output, $err>
+                >,
+            }
+        }
+
+        impl $future_name {
+            #[doc = concat!("Create a new `", stringify!($future_name), "` with the given future.")]
+            pub fn new<F>(future: F) -> Self
+            where
+                F: std::future::Future<Output = Result<$output, $err>> + Send + 'static,
+            {
+                Self {
+                    inner: aws_smithy_async::future::now_or_later::NowOrLater::new(Box::pin(future)),
+                }
+            }
+
+            #[doc = concat!("
+            Create a new `", stringify!($future_name), "` with the given boxed future.
+
+            Use this if you already have a boxed future to avoid double boxing it.
+            ")]
+            pub fn new_boxed(
+                future: std::pin::Pin<
+                    Box<dyn std::future::Future<Output = Result<$output, $err>> + Send>,
+                >,
+            ) -> Self {
+                Self {
+                    inner: aws_smithy_async::future::now_or_later::NowOrLater::new(future),
+                }
+            }
+
```

Comment:
can we use the downcast trick here?

Reply:
I don't think it would work so well here since the `Pin<Box<...>>` type would need to be just right, and the chances of that are slim.